### PR TITLE
Logging Module + Dynamic Buy/Sell Prop

### PR DESCRIPTION
**Pull request by jbdelmundo**
### Description

Added the following changes:
- Docstrings in `BaseStrategy`
- Added a new logging module instead of simply doing `print()` inside the `BaseStrategy`
- Logging inside the strategy can be accessed using `self.logging.log()`
- Separated the function of adding periodic cash
- `buy_signal()` and `sell_signal()` can now return `(signal, buy/sell_prop)` instead of just boolean (still backwards compatible)
- Added `trade_history`
- Set default signal to `False` (no action)

### Checklist
 - [ X] I am making a pull request from a branch other than master
 - [ X] I have read the CONTRIBUTING.md
 - [ X] I have added/edited documentation in a relevant docs/docusaurus markdown file
 - [X ] (For new docs, check if not applicable) I have added the id of a new docs md to the docs/docusaurus/sidebars.js file

### CODE REVIEW

**Comment by jbdelmundo** [comment]
This might be a large change so comments, requests for change are welcome! :)

**Review by rafmacalaba — COMMENTED** [review]
lgtm

**Review by rafmacalaba — APPROVED** [review]

**Comment by jgendrinal** [comment]
*has no idea what I'm doing* lgtm!! 😅 I'll defer to your judgement

File: python/fastquant/strategies/base.py [hunk 1]
```diff
@@ -30,14 +33,60 @@
     SELL_PROP,
     SHORT_MAX,
 )
+from fastquant.utils.strategy_logging import StrategyLogger
 
 
 class BaseStrategy(bt.Strategy):
     """
     Base Strategy template for all strategies to be added to fastquant
+
```
Comment (enzoampil) [inline python/fastquant/strategies/base.py:42]: These are beautiful docs, thanks so much for this!

File: python/fastquant/strategies/base.py [hunk 2]
```diff
@@ -30,14 +33,60 @@
...
+    add_cash_amount : int (default=0)
+        Added cash amount for periodic cash-ins. (can be used in cost averaging)
+    add_cash_frequency : str {`W` or `M`} (default=`W`)
+        Frequency of when to add cash (weekly or monthly)
+    channel: str (default=`""`)
+        Notification channel
+    symbol: str
+        Symbol of the asset traded
+
+    Renamed in code:
```
Comment (enzoampil) [inline python/fastquant/strategies/base.py:74]: Agree with the "internal" renaming, for increased clarity.

File: python/fastquant/strategies/base.py [hunk 3]
```diff
@@ -30,14 +33,60 @@
...
+        If set to `market`, excecutes the action (buy or sell) at current bar (or day),
+        If set to `close`, executees the action on the next available bar
+    allow_short : bool (default=False)
+        allow short selling
+    short_max : float
+        The max incremental short allowed is the short that would lead to a cumulative short position
+        equal to the maximum short position (initial cash times the maximum short ratio, which is 1.5 by default)
+    add_cash_amount : int (default=0)
+        Added cash amount for periodic cash-ins. (can be used in cost averaging)
+    add_cash_frequency : str {`W` or `M`} (default=`W`)
```
Comment (enzoampil) [inline python/fastquant/strategies/base.py:67]:
```suggestion
    add_cash_frequency : str {`W`, `M`, or custom cron notion} (default=`W`)
```

Comment (enzoampil) [reply python/fastquant/strategies/base.py:67]: Just added clarification that any custom cron notation can be used to indicate cash inflow frequency / timing

File: python/fastquant/strategies/base.py [hunk 4]
```diff
@@ -150,19 +162,25 @@ def __init__(self):
...
+        # Warn if user forgets to set a value to commission
+        if self.commission == 0:
+            warnings.warn(
+                "Comission value is set to 0. This might yield unrealistic results. Set using the `commission` paramter in `backtest()`",
+                UserWarning,
+            )
+
     def buy_signal(self):
-        return True
+        return False  # Safer to set default action to do nothing
```
Comment (enzoampil) [inline python/fastquant/strategies/base.py:173]: Looks good

File: python/fastquant/strategies/base.py [hunk 5]
```diff
@@ -30,14 +33,60 @@
...
+        buy proportion, percentage of available units to buy, setting to 1.0 buys all afforded units
+    sell_prop : float
+        sell proportion, percentage of owned units to sell
+    commission: float
+        Percentage of comission
+    stop_loss : float
+        stop loss percent
+    stop_trail : float
+        stop trail percent
+    execution_type : str {`market`,`close`} (default=`market`)
```
Comment (enzoampil) [inline python/fastquant/strategies/base.py:57]: @jbdelmundo Any specific reason why you decided to leave out the `open` option that was previously implemented? :smile:

File: python/fastquant/strategies/base.py [hunk 6]
```diff
@@ -172,261 +190,307 @@ def notify_order(self, order):
...
-                    )
-                    if max_position_size > 0:
-                        if self.transaction_logging:
-                            self.log("SELL CREATE, %.2f" % self.dataopen[1])
-                        self.order = self.sell(
-                            size=min(sell_prop_size, max_position_size)
-                        )
 
-            elif stock_value > 0:
+            # Buy on tomorrow's closing price
```
Comment (enzoampil) [inline python/fastquant/strategies/base.py:359]: @jbdelmundo  Yeah, I think best to still include the `open` type of execution. Technically, next day open is the most recommended approach to implementing the buy and sell; although, we use current / next day close as a proxy when next day open is not available.

File: python/fastquant/strategies/base.py [hunk 7]
```diff
@@ -172,261 +190,307 @@ def notify_order(self, order):
...
-                            exectype=bt.Order.Close,
-                        )
-                else:
-                    # Sell based on the opening price of the next closing day (only works "open" data exists in the dataset)
+                self.order = self.sell(
+                    size=min(sell_proportion_size, max_position_size)
+                )
+
+        # Buy based on the opening price of the next closing day (only works "open" data exists in the dataset)
+        else:
```
Comment (enzoampil) [inline python/fastquant/strategies/base.py:423]: @jbdelmundo I notice that for `create_buy_order` the execution types are `market`, and `close`, while for `create_short_sell_order`, you stuck to the original `close`, and `open` execution types. Is there a reason for why they're not consistent?

**Review by enzoampil — CHANGES_REQUESTED** [review]
Hey @jbdelmundo , thanks so much for this PR. `BaseStrategy` looks 5 times cleaner, and the buy / sell proportions look very interesting :smile:

Let's start with the initial comments that I just gave, most importantly regarding the new "execution types" being implemented, and the inconsistency between the types used for `create_buy_order` vs `create_sell_order`.

My recommendation is to include all three execution types for both:

1. `close` - next time step close
2. `open` - next time step open (this is the preferred `execution_type` though requires an `open` column)
3. `market` - current time step close

*Note: There's current handling that makes `open` the same as `market` in the case that `open` is chosen, but the column doesn't exist in the input dateframe.*

**Review by jgendrinal — APPROVED** [review]

File: python/fastquant/strategies/base.py (hunk 5 above)
Comment (jbdelmundo) [reply python/fastquant/strategies/base.py:57]: Actually I am having trouble making the backtest use the next opening price. 
I have tried using Market and Close orders in backtrader but it seems both use the closing price.
For Close orders, it uses the next day (or bar) close price, and for Market orders, it uses the current closing price

I am using the Ternary Strategy to manually provide buy and sell signals and review them on the order history

**Comment by jbdelmundo** [comment]
Reopened [here](https://github.com/enzoampil/fastquant/pull/380).
I was renaming my remote branch and got deleted

**Pull request closed without merging**